File: Source/Editor/ProjectInfo.cs
// Copyright (c) Wojciech Figat. All rights reserved.

using System;
using System.IO;
using System.Collections.Generic;
using FlaxEngine;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace FlaxEditor
{
    /// <summary>
    /// 
    /// </summary>
    public class FlaxVersionConverter : JsonConverter
    {
        // Original implementation is based on Newtonsoft.Json VersionConverter
        /// <summary>
        /// Writes the JSON representation of the object.
        /// </summary>
        /// <param name="writer">The <see cref="JsonWriter"/> to write to.</param>
        /// <param name="value">The value.</param>
        /// <param name="serializer">The calling serializer.</param>
        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
        {
            if (value == null)
                writer.WriteNull();
            else if (value is Version)
                writer.WriteValue(value.ToString());
            else
                throw new JsonSerializationException("Expected Version object value");
        }

        /// <summary>
        /// Reads the JSON representation of the object.
        /// </summary>
        /// <param name="reader">The <see cref="JsonReader"/> to read from.</param>
        /// <param name="objectType">Type of the object.</param>
        /// <param name="existingValue">The existing property value of the JSON that is being converted.</param>
        /// <param name="serializer">The calling serializer.</param>
        /// <returns>The object value.</returns>
        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
        {
            if (reader.TokenType == JsonToken.Null)
                return null;

            if (reader.TokenType == JsonToken.StartObject)
            {
                try
                {
                    reader.Read();
                    var values = new Dictionary<string, int>();
                    while (reader.TokenType == JsonToken.PropertyName)
                    {
                        var key = reader.Value as string;
                        reader.Read();
                        var val = (long)reader.Value;
                        reader.Read();
                        values.Add(key, (int)val);
                    }

                    values.TryGetValue("Major", out var major);
                    values.TryGetValue("Minor", out var minor);
                    if (!values.TryGetValue("Build", out var build))
                        build = -1;
                    if (!values.TryGetValue("Revision", out var revision))
                        revision = -1;

                    if (build <= 0)
                        return new Version(major, minor);
                    if (revision <= 0)
                        return new Version(major, minor, build);
                    return new Version(major, minor, build, revision);
                }
                catch (Exception ex)
                {
                    throw new Exception(String.Format("Error parsing version string: {0}", reader.Value), ex);
                }
            }
            if (reader.TokenType == JsonToken.String)
            {
                try
                {
                    return new Version((string)reader.Value!);
                }
                catch (Exception ex)
                {
                    throw new Exception(String.Format("Error parsing version string: {0}", reader.Value), ex);
                }
            }
            throw new Exception(String.Format("Unexpected token or value when parsing version. Token: {0}, Value: {1}", reader.TokenType, reader.Value));
        }

        /// <summary>
        /// Determines whether this instance can convert the specified object type.
        /// </summary>
        /// <param name="objectType">Type of the object.</param>
        /// <returns><c>true</c> if this instance can convert the specified object type; otherwise, <c>false</c>.</returns>
        public override bool CanConvert(Type objectType)
        {
            return objectType == typeof(Version);
        }
    }

    /// <summary>
    /// Serializes a project default scene as a persistent asset object identifier.
    /// </summary>
    public sealed class ProjectDefaultSceneConverter : JsonConverter
    {
        /// <inheritdoc />
        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
        {
            if (!(value is Guid id) || id == Guid.Empty)
                throw new JsonSerializationException("Expected a valid default scene GUID.");
            writer.WriteValue(id.ToString());
        }

        /// <inheritdoc />
        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
        {
            if (reader.TokenType == JsonToken.String && Guid.TryParseExact((string)reader.Value, "N", out var guid) && guid != Guid.Empty)
                return guid;
            throw new JsonSerializationException("DefaultScene must contain a valid GUID string.");
        }

        /// <inheritdoc />
        public override bool CanConvert(Type objectType)
        {
            return objectType == typeof(Guid);
        }
    }

    /// <summary>
    /// Contains information about Flax project.
    /// </summary>
    public sealed class ProjectInfo
    {
        private static List<ProjectInfo> _projectsCache;

        /// <summary>
        /// The project reference.
        /// </summary>
        public class Reference
        {
            /// <summary>
            /// The referenced project name.
            /// </summary>
            public string Name;

            /// <summary>
            /// The referenced project.
            /// </summary>
            [NonSerialized]
            public ProjectInfo Project;

            /// <inheritdoc />
            public override string ToString()
            {
                return Name;
            }
        }

        /// <summary>
        /// The project name.
        /// </summary>
        public string Name;

        /// <summary>
        /// The project file path.
        /// </summary>
        [NonSerialized]
        public string ProjectPath;

        /// <summary>
        /// The project root folder path.
        /// </summary>
        [NonSerialized]
        public string ProjectFolderPath;

        /// <summary>
        /// The project version.
        /// </summary>
        public Version Version;

        /// <summary>
        /// The source asset system format required by this project.
        /// </summary>
        public int AssetSystemVersion;

        /// <summary>
        /// The source GUID of Content/Settings/ProjectSettingsIndex.settings.
        /// </summary>
        public string ProjectSettingsIndexGuid;

        /// <summary>
        /// The project publisher company.
        /// </summary>
        public string Company = string.Empty;

        /// <summary>
        /// The project copyright note.
        /// </summary>
        public string Copyright = string.Empty;

        /// <summary>
        /// The name of the build target to use for the game building (final, cooked game code).
        /// </summary>
        public string GameTarget;

        /// <summary>
        /// The name of the build target to use for the game in editor building (editor game code).
        /// </summary>
        public string EditorTarget;

        /// <summary>
        /// The project references.
        /// </summary>
        public Reference[] References = new Reference[0];

        /// <summary>
        /// The default scene asset identifier to open on project startup.
        /// </summary>
        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
        [JsonConverter(typeof(ProjectDefaultSceneConverter))]
        public Guid DefaultScene;

        /// <summary>
        /// The default scene spawn point (position and view direction).
        /// </summary>
        public Ray DefaultSceneSpawn;

        /// <summary>
        /// The minimum version supported by this project.
        /// </summary>
        public Version MinEngineVersion;

        /// <summary>
        /// The user-friendly nickname of the engine installation to use when opening the project. Can be used to open game project with a custom engine distributed for team members. This value must be the same in engine and game projects to be paired.
        /// </summary>
        public string EngineNickname;

        /// <summary>
        /// Gets all projects including this project, it's references and their references (any deep level of references).
        /// </summary>
        /// <returns>The collection of projects.</returns>
        public HashSet<ProjectInfo> GetAllProjects()
        {
            var result = new HashSet<ProjectInfo>();
            GetAllProjects(result);
            return result;
        }

        private void GetAllProjects(HashSet<ProjectInfo> result)
        {
            result.Add(this);
            foreach (var reference in References)
                reference.Project.GetAllProjects(result);
        }

        /// <summary>
        /// Saves the project file.
        /// </summary>
        public void Save()
        {
            var contents = FlaxEngine.Json.JsonSerializer.Serialize(this);
            File.WriteAllText(ProjectPath, contents);
        }

        private static void ValidateCurrentFormat(string contents)
        {
            const string guidance = "Unsupported current project format. Run the separate offline migrator from the old branch before opening this project.";
            JObject root;
            using (var reader = new JsonTextReader(new StringReader(contents)))
            {
                root = JObject.Load(reader, new JsonLoadSettings
                {
                    CommentHandling = CommentHandling.Ignore,
                    DuplicatePropertyNameHandling = DuplicatePropertyNameHandling.Error,
                });
            }
            bool StringField(string name, bool allowEmpty = false) =>
                root[name]?.Type == JTokenType.String && (allowEmpty || !string.IsNullOrEmpty(root.Value<string>(name)));
            if (!StringField("Name") || !StringField("Version") || !StringField("Company", true) ||
                !StringField("Copyright", true) || !StringField("GameTarget") || !StringField("EditorTarget") ||
                !StringField("MinEngineVersion") || !System.Version.TryParse(root.Value<string>("Version"), out _) ||
                !System.Version.TryParse(root.Value<string>("MinEngineVersion"), out _) ||
                root["AssetSystemVersion"]?.Type != JTokenType.Integer || root.Value<int>("AssetSystemVersion") != 2)
                throw new InvalidDataException(guidance);
            var settingsGuidText = root.Value<string>("ProjectSettingsIndexGuid");
            if (!StringField("ProjectSettingsIndexGuid") || !Guid.TryParseExact(settingsGuidText, "N", out var settingsGuid) ||
                settingsGuid == Guid.Empty || settingsGuid.ToString("N") != settingsGuidText)
                throw new InvalidDataException(guidance);
            if (!(root["References"] is JArray references))
                throw new InvalidDataException(guidance);
            foreach (var reference in references)
            {
                if (!(reference is JObject referenceObject) || referenceObject["Name"]?.Type != JTokenType.String ||
                    string.IsNullOrEmpty(referenceObject.Value<string>("Name")))
                    throw new InvalidDataException(guidance);
            }
            if (root.TryGetValue("DefaultScene", out var defaultScene))
            {
                if (defaultScene.Type != JTokenType.String ||
                    !Guid.TryParseExact(defaultScene.Value<string>(), "N", out var guid) || guid == Guid.Empty ||
                    guid.ToString("N") != defaultScene.Value<string>())
                    throw new InvalidDataException(guidance);
            }
            if (root.TryGetValue("EngineNickname", out var nickname) &&
                (nickname.Type != JTokenType.String || string.IsNullOrEmpty(nickname.Value<string>())))
                throw new InvalidDataException(guidance);
            if (root.TryGetValue("DefaultSceneSpawn", out var spawn))
            {
                bool Coordinate(JObject vector, string name) =>
                    vector[name]?.Type == JTokenType.Float || vector[name]?.Type == JTokenType.Integer;
                if (!(spawn is JObject ray) || !(ray["Position"] is JObject position) || !(ray["Direction"] is JObject direction) ||
                    !Coordinate(position, "X") || !Coordinate(position, "Y") || !Coordinate(position, "Z") ||
                    !Coordinate(direction, "X") || !Coordinate(direction, "Y") || !Coordinate(direction, "Z"))
                    throw new InvalidDataException(guidance);
            }
            if (root.TryGetValue("Configuration", out var configuration) && configuration.Type != JTokenType.Object)
                throw new InvalidDataException(guidance);
        }

        /// <summary>
        /// Loads the project from the specified file.
        /// </summary>
        /// <param name="path">The path.</param>
        /// <returns>The loaded project.</returns>
        public static ProjectInfo Load(string path)
        {
            // Try to reuse loaded file
            path = StringUtils.RemovePathRelativeParts(path);
            if (_projectsCache == null)
                _projectsCache = new List<ProjectInfo>();
            for (int i = 0; i < _projectsCache.Count; i++)
            {
                if (_projectsCache[i].ProjectPath == path)
                    return _projectsCache[i];
            }

            Profiler.BeginEvent(path);
            try
            {
                // Load
                var contents = File.ReadAllText(path);
                ValidateCurrentFormat(contents);
                var project = JsonConvert.DeserializeObject<ProjectInfo>(contents, new JsonSerializerSettings() { Converters = new[] { new FlaxVersionConverter() } });
                project.ProjectPath = path;
                project.ProjectFolderPath = StringUtils.NormalizePath(Path.GetDirectoryName(path));

                // Process project data
                if (string.IsNullOrEmpty(project.Name))
                    throw new Exception("Missing project name.");
                if (project.Version.Revision == 0)
                    project.Version = new Version(project.Version.Major, project.Version.Minor, project.Version.Build);
                if (project.Version.Build == 0 && project.Version.Revision == -1)
                    project.Version = new Version(project.Version.Major, project.Version.Minor);
                foreach (var reference in project.References)
                {
                    string referencePath;
                    if (reference.Name.StartsWith("$(EnginePath)"))
                    {
                        // Relative to engine root
                        referencePath = Path.Combine(Globals.StartupFolder, reference.Name.Substring(14));
                    }
                    else if (reference.Name.StartsWith("$(ProjectPath)"))
                    {
                        // Relative to project root
                        referencePath = Path.Combine(project.ProjectFolderPath, reference.Name.Substring(15));
                    }
                    else if (Path.IsPathRooted(reference.Name))
                    {
                        // Relative to workspace
                        referencePath = Path.Combine(Environment.CurrentDirectory, reference.Name);
                    }
                    else
                    {
                        // Absolute
                        referencePath = reference.Name;
                    }

                    // Load referenced project
                    reference.Project = Load(referencePath);
                }

                // Project loaded
                Editor.Log($"Loaded project {project.Name}, version {project.Version}");
                _projectsCache.Add(project);
                return project;
            }
            catch
            {
                // Failed to load project
                Editor.LogError("Failed to load project \"" + path + "\".");
                throw;
            }
            finally
            {
                Profiler.EndEvent();
            }
        }

        /// <inheritdoc />
        public override string ToString()
        {
            return $"{Name} ({ProjectPath})";
        }
    }
}

File: Source/Editor/ProjectInfo.cpp
// Copyright (c) Wojciech Figat. All rights reserved.

#include "ProjectInfo.h"
#include "Engine/Platform/FileSystem.h"
#include "Engine/Platform/File.h"
#include "Engine/Core/Log.h"
#include "Engine/Engine/Globals.h"
#include "Engine/Core/Math/Quaternion.h"
#include "Engine/Profiler/ProfilerMemory.h"
#include "Engine/Serialization/JsonWriters.h"
#include "Engine/Serialization/JsonTools.h"
#include "Engine/Content/Documents/CanonicalJsonWriter.h"
#include <ThirdParty/pugixml/pugixml.hpp>
using namespace pugi;

Array<ProjectInfo*> ProjectInfo::ProjectsCache;

struct XmlCharAsChar
{
#if PLATFORM_TEXT_IS_CHAR16
    Char* Str = nullptr;
    
    XmlCharAsChar(const pugi::char_t* str)
    {
        if (!str)
            return;
        int32 length = 0;
        while (str[length])
            length++;
        Str = (Char*)Platform::Allocate(length * sizeof(Char), sizeof(Char));
        for (int32 i = 0; i <= length; i++)
            Str[i] = (Char)str[i];
    }

    ~XmlCharAsChar()
    {
        Platform::Free(Str);
    }
#else
    const Char* Str;

    XmlCharAsChar(const pugi::char_t* str)
        : Str(str)
    {
    }
#endif
};

void ShowProjectLoadError(const Char* errorMsg, const String& projectRootFolder)
{
    Platform::Error(String::Format(TEXT("Failed to load project. {0}\nPath: '{1}'"), errorMsg, projectRootFolder));
}

Vector3 GetVector3FromXml(const xml_node& parent, const PUGIXML_CHAR* name, const Vector3& defaultValue)
{
    const auto node = parent.child(name);
    if (!node.empty())
    {
        const auto x = node.child_value(PUGIXML_TEXT("X"));
        const auto y = node.child_value(PUGIXML_TEXT("Y"));
        const auto z = node.child_value(PUGIXML_TEXT("Z"));
        if (x && y && z)
        {
            XmlCharAsChar xs(x), ys(y), zs(z);
            Float3 v;
            if (!StringUtils::Parse(xs.Str, &v.X) && !StringUtils::Parse(ys.Str, &v.Y) && !StringUtils::Parse(zs.Str, &v.Z))
            {
                return (Vector3)v;
            }
        }
    }
    return defaultValue;
}

int32 GetIntFromXml(const xml_node& parent, const PUGIXML_CHAR* name, const int32 defaultValue)
{
    const auto node = parent.child_value(name);
    if (node)
    {
        XmlCharAsChar s(node);
        int32 v;
        if (!StringUtils::Parse(s.Str, &v))
        {
            return v;
        }
    }

    return defaultValue;
}

bool ProjectInfo::SaveProject()
{
    // Serialize object to Json
    rapidjson_flax::StringBuffer buffer;
    PrettyJsonWriter writerObj(buffer);
    auto& stream = *(JsonWriter*)&writerObj;
    stream.StartObject();
    {
        stream.JKEY("Name");
        stream.String(Name);

        stream.JKEY("Version");
        stream.String(Version.ToString());

        stream.JKEY("AssetSystemVersion");
        stream.Int(AssetSystemVersion);

        stream.JKEY("ProjectSettingsIndexGuid");
        stream.Guid(ProjectSettingsIndexGuid);

        stream.JKEY("Company");
        stream.String(Company);

        stream.JKEY("Copyright");
        stream.String(Copyright);

        stream.JKEY("GameTarget");
        stream.String(GameTarget);

        stream.JKEY("EditorTarget");
        stream.String(EditorTarget);

        stream.JKEY("References");
        stream.StartArray();
        for (auto& reference : References)
        {
            stream.StartObject();
            stream.JKEY("Name");
            stream.String(reference.Name);
            stream.EndObject();
        }
        stream.EndArray();

        if (DefaultScene.IsValid())
        {
            stream.JKEY("DefaultScene");
            stream.Guid(DefaultScene);
        }

        if (DefaultSceneSpawn != Ray(Vector3::Zero, Vector3::Forward))
        {
            stream.JKEY("DefaultSceneSpawn");
            stream.Ray(DefaultSceneSpawn);
        }

        stream.JKEY("MinEngineVersion");
        stream.String(MinEngineVersion.ToString());

        if (EngineNickname.HasChars())
        {
            stream.JKEY("EngineNickname");
            stream.String(EngineNickname);
        }
    }
    stream.EndObject();

    // Write to file
    return File::WriteAllBytes(ProjectPath, (const byte*)buffer.GetString(), (int32)buffer.GetSize());
}

bool ProjectInfo::LoadProject(const String& projectPath)
{
    // Load Json file
    StringAnsi fileData;
    if (File::ReadAllText(projectPath, fileData))
    {
        ShowProjectLoadError(TEXT("Failed to read file contents."), projectPath);
        return true;
    }

    // Parse Json data
    rapidjson_flax::Document document;
    document.Parse(fileData.Get(), fileData.Length());
    if (document.HasParseError())
    {
        ShowProjectLoadError(TEXT("Failed to parse project contents. Ensure to have valid Json format."), projectPath);
        return true;
    }

    const auto unsupported = [&projectPath](const Char* reason)
    {
        const String message = String::Format(TEXT("Unsupported current project format: {0} Run the separate offline migrator from the old branch before opening this project."), reason);
        ShowProjectLoadError(*message, projectPath);
        return true;
    };
    CanonicalJsonError canonicalError;
    if (!document.IsObject() || CanonicalJsonWriter::Validate(document, canonicalError))
        return unsupported(TEXT("the manifest is not a unique-key JSON object."));
    const auto requireString = [&document](const char* name, bool allowEmpty = false)
    {
        const auto member = document.FindMember(name);
        return member != document.MemberEnd() && member->value.IsString() &&
            (allowEmpty || member->value.GetStringLength() != 0);
    };
    if (!requireString("Name") || !requireString("Version") || !requireString("Company", true) ||
        !requireString("Copyright", true) || !requireString("GameTarget") || !requireString("EditorTarget") ||
        !requireString("MinEngineVersion"))
        return unsupported(TEXT("a required current string field is missing or invalid."));
    const auto assetSystemVersion = document.FindMember("AssetSystemVersion");
    if (assetSystemVersion == document.MemberEnd() || !assetSystemVersion->value.IsInt() || assetSystemVersion->value.GetInt() != 2)
        return unsupported(TEXT("AssetSystemVersion must be exactly 2."));
    const auto settingsGuid = document.FindMember("ProjectSettingsIndexGuid");
    Guid parsedSettingsGuid;
    if (settingsGuid == document.MemberEnd() || !settingsGuid->value.IsString() || settingsGuid->value.GetStringLength() != 32 ||
        Guid::Parse(StringAnsiView(settingsGuid->value.GetString(), settingsGuid->value.GetStringLength()), parsedSettingsGuid) ||
        !parsedSettingsGuid.IsValid() || StringAnsi(parsedSettingsGuid.ToString(Guid::FormatType::N).ToLower()) !=
            StringAnsiView(settingsGuid->value.GetString(), settingsGuid->value.GetStringLength()))
        return unsupported(TEXT("ProjectSettingsIndexGuid must be a lowercase 32-hex GUID."));
    ::Version parsedVersion;
    const auto versionText = document.FindMember("Version");
    const auto minimumText = document.FindMember("MinEngineVersion");
    if (::Version::Parse(String(StringAnsiView(versionText->value.GetString(), versionText->value.GetStringLength())), &parsedVersion) ||
        ::Version::Parse(String(StringAnsiView(minimumText->value.GetString(), minimumText->value.GetStringLength())), &parsedVersion))
        return unsupported(TEXT("Version and MinEngineVersion must be parseable strings."));
    const auto referencesValue = document.FindMember("References");
    if (referencesValue == document.MemberEnd() || !referencesValue->value.IsArray())
        return unsupported(TEXT("References must be an array."));
    for (const auto& reference : referencesValue->value.GetArray())
    {
        if (!reference.IsObject())
            return unsupported(TEXT("every project reference requires a nonempty Name."));
        const auto name = reference.FindMember("Name");
        if (name == reference.MemberEnd() || !name->value.IsString() || name->value.GetStringLength() == 0)
            return unsupported(TEXT("every project reference requires a nonempty Name."));
    }
    const auto defaultSceneValue = document.FindMember("DefaultScene");
    if (defaultSceneValue != document.MemberEnd())
    {
        if (!defaultSceneValue->value.IsString())
            return unsupported(TEXT("DefaultScene must use the current GUID-only form."));
        Guid parsedGuid;
        if (defaultSceneValue->value.GetStringLength() != 32 ||
            Guid::Parse(StringAnsiView(defaultSceneValue->value.GetString(), defaultSceneValue->value.GetStringLength()), parsedGuid) || !parsedGuid.IsValid() ||
            StringAnsi(parsedGuid.ToString(Guid::FormatType::N).ToLower()) != StringAnsiView(defaultSceneValue->value.GetString(), defaultSceneValue->value.GetStringLength()))
            return unsupported(TEXT("DefaultScene must contain one canonical GUID."));
    }
    const auto engineNickname = document.FindMember("EngineNickname");
    if (engineNickname != document.MemberEnd() && (!engineNickname->value.IsString() || engineNickname->value.GetStringLength() == 0))
        return unsupported(TEXT("EngineNickname must be a nonempty string when present."));
    const auto defaultSpawn = document.FindMember("DefaultSceneSpawn");
    if (defaultSpawn != document.MemberEnd())
    {
        const auto validVector = [](const rapidjson_flax::Value& value)
        {
            if (!value.IsObject())
                return false;
            const auto x = value.FindMember("X");
            const auto y = value.FindMember("Y");
            const auto z = value.FindMember("Z");
            return x != value.MemberEnd() && x->value.IsNumber() && y != value.MemberEnd() && y->value.IsNumber() &&
                z != value.MemberEnd() && z->value.IsNumber();
        };
        if (!defaultSpawn->value.IsObject())
            return unsupported(TEXT("DefaultSceneSpawn must use the current Ray object form."));
        const auto position = defaultSpawn->value.FindMember("Position");
        const auto direction = defaultSpawn->value.FindMember("Direction");
        if (position == defaultSpawn->value.MemberEnd() || direction == defaultSpawn->value.MemberEnd() ||
            !validVector(position->value) || !validVector(direction->value))
            return unsupported(TEXT("DefaultSceneSpawn must use the current Ray object form."));
    }
    const auto configuration = document.FindMember("Configuration");
    if (configuration != document.MemberEnd() && !configuration->value.IsObject())
        return unsupported(TEXT("Configuration must be an object when present."));

    // Parse properties
    Name = JsonTools::GetString(document, "Name", String::Empty);
    ProjectPath = projectPath;
    ProjectFolderPath = StringUtils::GetDirectoryName(projectPath);
    const auto versionMember = document.FindMember("Version");
    if (versionMember != document.MemberEnd())
    {
        auto& version = versionMember->value;
        if (version.IsString())
        {
            Version::Parse(version.GetText(), &Version);
        }
    }
    if (Version.Revision() == 0)
        Version = ::Version(Version.Major(), Version.Minor(), Version.Build());
    if (Version.Build() == 0 && Version.Revision() == -1)
        Version = ::Version(Version.Major(), Version.Minor());
    AssetSystemVersion = JsonTools::GetInt(document, "AssetSystemVersion", 0);
    ProjectSettingsIndexGuid = JsonTools::GetGuid(document, "ProjectSettingsIndexGuid");
    Company = JsonTools::GetString(document, "Company", String::Empty);
    Copyright = JsonTools::GetString(document, "Copyright", String::Empty);
    GameTarget = JsonTools::GetString(document, "GameTarget", String::Empty);
    EditorTarget = JsonTools::GetString(document, "EditorTarget", String::Empty);
    EngineNickname = JsonTools::GetString(document, "EngineNickname", String::Empty);
    const auto referencesMember = document.FindMember("References");
    if (referencesMember != document.MemberEnd())
    {
        const auto& references = referencesMember->value.GetArray();
        References.Resize(references.Size());
        for (int32 i = 0; i < References.Count(); i++)
        {
            auto& reference = References[i];
            auto& value = references[i];
            reference.Name = JsonTools::GetString(value, "Name", String::Empty);

            String referencePath;
            if (reference.Name.StartsWith(TEXT("$(EnginePath)")))
            {
                // Relative to engine root
                referencePath = Globals::StartupFolder / reference.Name.Substring(14);
            }
            else if (reference.Name.StartsWith(TEXT("$(ProjectPath)")))
            {
                // Relative to project root
                referencePath = ProjectFolderPath / reference.Name.Substring(15);
            }
            else if (FileSystem::IsRelative(reference.Name))
            {
                // Relative to workspace
                referencePath = Globals::StartupFolder / reference.Name;
            }
            else
            {
                // Absolute
                referencePath = reference.Name;
            }
            StringUtils::PathRemoveRelativeParts(referencePath);

            // Load referenced project
            reference.Project = Load(referencePath);
            if (reference.Project == nullptr)
            {
                LOG(Error, "Failed to load referenced project ({0}, from {1})", reference.Name, referencePath);
                return true;
            }
        }
    }
    DefaultScene = Guid::Empty;
    const auto defaultSceneMember = document.FindMember("DefaultScene");
    if (defaultSceneMember != document.MemberEnd())
    {
        const auto& value = defaultSceneMember->value;
        if (value.IsString() && value.GetStringLength() == 32)
        {
            Guid guid;
            if (Guid::Parse(StringAnsiView(value.GetString(), value.GetStringLength()), guid) || !guid.IsValid())
            {
                ShowProjectLoadError(TEXT("Invalid DefaultScene GUID."), projectPath);
                return true;
            }
            DefaultScene = guid;
        }
        else
        {
            ShowProjectLoadError(TEXT("DefaultScene must contain one asset GUID."), projectPath);
            return true;
        }
    }
    DefaultSceneSpawn = JsonTools::GetRay(document, "DefaultSceneSpawn", Ray(Vector3::Zero, Vector3::Forward));
    const auto minEngineVersionMember = document.FindMember("MinEngineVersion");
    if (minEngineVersionMember != document.MemberEnd())
    {
        auto& minEngineVersion = minEngineVersionMember->value;
        if (minEngineVersion.IsString())
        {
            Version::Parse(minEngineVersion.GetText(), &MinEngineVersion);
        }
    }

    // Validate properties
    if (Name.Length() == 0)
    {
        ShowProjectLoadError(TEXT("Missing project name."), projectPath);
        return true;
    }

    return false;
}

bool ProjectInfo::LoadOldProject(const String& projectPath)
{
    // Open Xml file
    xml_document doc;
    const xml_parse_result result = doc.load_file((const PUGIXML_CHAR*)*projectPath);
    if (result.status)
    {
        ShowProjectLoadError(TEXT("Xml file parsing error."), projectPath);
        return true;
    }

    // Get root node
    const xml_node root = doc.child(PUGIXML_TEXT("Project"));
    if (!root)
    {
        ShowProjectLoadError(TEXT("Missing Project root node in xml file."), projectPath);
        return true;
    }

    // Load data
    Name = (const Char*)root.child_value(PUGIXML_TEXT("Name"));
    ProjectPath = projectPath;
    ProjectFolderPath = StringUtils::GetDirectoryName(projectPath);
    DefaultScene = Guid::Empty;
    const auto defaultScene = root.child_value(PUGIXML_TEXT("DefaultSceneId"));
    if (defaultScene)
    {
        Guid legacyGuid;
        if (!Guid::Parse((const Char*)defaultScene, legacyGuid))
            DefaultScene = legacyGuid;
    }
    DefaultSceneSpawn.Position = GetVector3FromXml(root, PUGIXML_TEXT("DefaultSceneSpawnPos"), Vector3::Zero);
    DefaultSceneSpawn.Direction = Quaternion::Euler(GetVector3FromXml(root, PUGIXML_TEXT("DefaultSceneSpawnDir"), Vector3::Zero)) * Vector3::Forward;
    MinEngineVersion = ::Version(0, 0, GetIntFromXml(root, PUGIXML_TEXT("MinVersionSupported"), 0));

    // Always reference engine project
    auto& flaxReference = References.AddOne();
    flaxReference.Name = TEXT("$(EnginePath)/Flax.flaxproj");
    flaxReference.Project = Load(Globals::StartupFolder / TEXT("Flax.flaxproj"));
    if (!flaxReference.Project)
    {
        ShowProjectLoadError(TEXT("Failed to load Flax Engine project."), projectPath);
        return true;
    }

    return false;
}

ProjectInfo* ProjectInfo::Load(const String& path)
{
    // Try to reuse loaded file
    for (int32 i = 0; i < ProjectsCache.Count(); i++)
    {
        if (ProjectsCache[i]->ProjectPath == path)
            return ProjectsCache[i];
    }

    // Load
    PROFILE_MEM(Editor);
    auto project = New<ProjectInfo>();
    if (project->LoadProject(path))
    {
        Delete(project);
        return nullptr;
    }

    // Cache project
    ProjectsCache.Add(project);
    return project;
}
